File: pubnub-java/src/main/java/com/pubnub/internal/endpoints/objects_api/members/GetChannelMembersImpl.java
package com.pubnub.internal.endpoints.objects_api.members;

import com.pubnub.api.endpoints.objects_api.members.GetChannelMembers;
import com.pubnub.api.endpoints.objects_api.utils.Include;
import com.pubnub.api.endpoints.objects_api.utils.PNSortKey;
import com.pubnub.api.endpoints.remoteaction.ExtendedRemoteAction;
import com.pubnub.api.endpoints.remoteaction.MappingRemoteAction;
import com.pubnub.api.models.consumer.objects.PNPage;
import com.pubnub.api.models.consumer.objects_api.member.PNGetChannelMembersResult;
import com.pubnub.internal.InternalPubNubClient;
import com.pubnub.internal.endpoints.DelegatingEndpoint;
import lombok.AllArgsConstructor;
import lombok.Setter;
import lombok.experimental.Accessors;

import java.util.Collection;
import java.util.Collections;

@Setter
@Accessors(chain = true, fluent = true)
public class GetChannelMembersImpl extends DelegatingEndpoint<PNGetChannelMembersResult> implements GetChannelMembers {

    private final String channel;
    private Integer limit = null;
    private PNPage page;
    private String filter;
    private Collection<PNSortKey> sort = Collections.emptyList();
    private boolean includeTotalCount;
    private boolean includeCustom;
    private Include.PNUUIDDetailsLevel includeUUID;

    public GetChannelMembersImpl(String channel, final InternalPubNubClient pubnubInstance) {
        super(pubnubInstance);
        this.channel = channel;
    }

    @Override
    protected ExtendedRemoteAction<PNGetChannelMembersResult> createAction() {
        return new MappingRemoteAction<>(pubnub.getChannelMembers(
                channel,
                limit,
                page,
                filter,
                SetChannelMembersImpl.toInternal(sort),
                includeTotalCount,
                includeCustom,
                SetChannelMembersImpl.toInternal(includeUUID)
        ), PNGetChannelMembersResult::from);
    }

    @AllArgsConstructor
    public static class Builder implements GetChannelMembers.Builder {
        private final InternalPubNubClient pubnubInstance;

        @Override
        public GetChannelMembers channel(final String channel) {
            return new GetChannelMembersImpl(channel, pubnubInstance);
        }
    }
}

File: pubnub-kotlin/src/main/kotlin/com/pubnub/internal/kotlin/endpoints/objects/member/GetChannelMembersImpl.kt
package com.pubnub.internal.kotlin.endpoints.objects.member

import com.pubnub.api.DelegatingEndpoint
import com.pubnub.api.endpoints.remoteaction.ExtendedRemoteAction
import com.pubnub.api.endpoints.remoteaction.MappingRemoteAction
import com.pubnub.internal.PubNubImpl
import com.pubnub.internal.endpoints.objects.member.GetChannelMembers
import com.pubnub.internal.endpoints.objects.member.IGetChannelMembers
import com.pubnub.internal.models.consumer.objects.member.PNMemberArrayResult

/**
 * @see [PubNubImpl.getChannelMembers]
 */
class GetChannelMembersImpl internal constructor(getChannelMembers: GetChannelMembers) :
    DelegatingEndpoint<com.pubnub.api.models.consumer.objects.member.PNMemberArrayResult, PNMemberArrayResult>(
        getChannelMembers
    ),
    IGetChannelMembers by getChannelMembers,
    com.pubnub.api.endpoints.objects.member.GetChannelMembers {
    override fun convertAction(remoteAction: ExtendedRemoteAction<PNMemberArrayResult>): ExtendedRemoteAction<com.pubnub.api.models.consumer.objects.member.PNMemberArrayResult> {
        return MappingRemoteAction(
            remoteAction,
            com.pubnub.api.models.consumer.objects.member.PNMemberArrayResult::from
        )
    }
}
